print actual template errors on $stderr

foxxx0 opened the pull request:
Previously the `EvaluatingParser` would implicitly call its own
`#assert_and_report()` method, which in turn hands it over to the
`IssueReporter#assert_and_report()` that tries to utilize puppet's
internal logging but in our rake task this is not properly setup and
thus the logged errors get lost.

The result just leaves the validation summary output, e.g.:
```
Language validation logged 2 errors. Giving up (file: templates/configs/jenkins_casc.sh.epp)
```

Which obviously isn't very helpful to the user.
This commit reworks the epp parsing and validation rake task, so that it
instead prints out a nice summary on `$stderr` that contains all
necessary information, e.g.:
```
ERRORS:
templates/configs/jenkins_casc.sh.epp:1:17: Illegal parameter name. The given name 'CASC_JENKINS_CONFIG' does not conform to the naming rule /^[a-z_]\w*$/
templates/configs/jenkins_casc.sh.epp:4:33: Illegal variable name, The given name 'CASC_JENKINS_CONFIG' does not conform to the naming rule /^((::)?[a-z]\w*)*((::)?[a-z_]\w*)$/
```

hlindberg commented on lib/puppet-syntax/tasks/puppet-syntax.rb:
```diff
@@ -49,8 +49,11 @@ def initialize(*args)
           errors = c.check(filelist_templates)
-          $stdout.puts "#{errors.join("\n")}\n" unless errors.empty?
-          exit 1 unless errors.empty?
+          unless errors.empty?
```
You probably also want output of any warnings

foxxx0 replied:
Good point, thanks!
done.

foxxx0 commented:
great point, added `STDOUT.puts()` for the warnings as well.

also parser exceptions are not properly rescued and added to the `errors[]` array, which make `rake spec` pass again.

ekohl commented on lib/puppet-syntax/templates.rb:
```diff
@@ -29,11 +29,33 @@ def check(filelist)
+          file = warn.file
+          line = warn.source_pos.line
+          column = warn.source_pos.pos
+          STDOUT.puts("WARNING: #{file}:#{line}:#{column}: #{message}")
```
It would be nice if this method didn't do anything with stdout. Perhaps it should return an `OpenStruct` with an array of warnings and an array of errors.

foxxx0 replied:
The template checks are now returning a ruby `Hash` with `:warnings` and `:errors` keys, each being an `Array`.
The `STDOUT`/`STDERR` output is now only done in the rake task again.

hlindberg commented:
Good work on this! I think there is one more case to consider wrt. exceptions - I think there may be a LexError as well if the template does not even lex ok - for example a malformed number, string that isn't closed etc.

foxxx0 commented:
There is a generic "catchall" rescue here: https://github.com/voxpupuli/puppet-syntax/pull/125/files#diff-a8986ba30354c68e43b0dc3441e884a2R61 which should take care of that.
